mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -197,63 +168,23 @@ impl PutObjectRequest for S3PutObjectRequest {
         let elapsed = self.start_time.elapsed();
         emit_throughput_metric(self.total_bytes, elapsed, "put_object");
 
-        check_response_headers(
-            self.response_headers
-                .lock()
-                .expect("must be able to acquire headers lock")
+        let response_headers_lock = self
+            .response_headers
+            .lock()
+            .expect("must be able to acquire headers lock");
+        let response_headers = response_headers_lock
+            .as_ref()
+            .expect("PUT response headers must be available at this point");
+        let try_get_header_value = |headers: &Headers, key: &str| -> Option<String> {
+            headers
+                .get(key)
+                .ok()
```

Comment:
nit: prefer a function if you are not closing over anything. Also, can be simplified a bit:
```suggestion
        fn try_get_header_value(headers: &Headers, key: &str) -> Option<String> {
            headers.get(key).ok()?.value().clone().into_string().ok()
        }
```